enhance unit tests of advance audit feature

CaoShuFeng opened the pull request:
This change addresses comments from @crassirostris 
https://github.com/kubernetes/kubernetes/pull/49115#discussion_r133416080

It does three things:
1. use auditinternal for unit test in filter stage
2. add a seperate unit test for Audit-ID http header
3. add unit test for audit log backend


**Release note**:
```
NONE
```

CaoShuFeng commented on staging/src/k8s.io/apiserver/plugin/pkg/audit/log/backend_test.go:
```diff
@@ -0,0 +1,176 @@
+			SourceIPs: []string{
+				"127.0.0.1",
+			},
+			// Uncomment the next line this test would fail:
```
Please take a look at here.
I think this is OK, and it's not an issue for me.

CaoShuFeng replied:
Please take a look at here.
I think this is OK, and it's not an issue for me.

crassirostris replied:
time.Truncate?

CaoShuFeng replied:
Yeap, time.Truncate runs happliy.

liggitt commented:
/unassign

CaoShuFeng commented:
/test pull-kubernetes-e2e-gce-etcd3

crassirostris commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit_test.go:
```diff
@@ -984,3 +704,71 @@ func TestAuditLevelNone(t *testing.T) {
+		resp := w.Result()
+		if test.expectedHeader {
+			if resp.Header.Get("Audit-ID") == "" {
+				t.Errorf("[%s] expected Audit-ID http header returned, but not returned", test.desc)
```
Instead of putting [%s] in front of each string, you can use the following approach:

```
for _, tc := range tcs {
  t.Run(tc.desc, func() {
    ...
  })
}
```

Then you can use `t.Failf` to fail the test without continuation

CaoShuFeng replied:
> Then you can use t.Failf to fail the test without continuation
Done.

crassirostris replied:
@CaoShuFeng Haven't pushed the changes?

CaoShuFeng replied:
>@CaoShuFeng Haven't pushed the changes?

I mean this is done in line 621.
I other places, we can use t.Errorf() to continue the tests to get more error messages. What do you think?

crassirostris replied:
But this particular place has `continue` afterwards :)

Anyway, if there's no kubernetes-wide policy about tests, do as you wish, that's just my feeling that writing t.Run once is better than writing it in each log line, at least for new tests.

crassirostris reviewed (commented):
LGTM, couple of nits

CaoShuFeng commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit_test.go:
```diff
@@ -874,30 +606,23 @@ func TestAuditJson(t *testing.T) {
-		if len(line) != len(test.expected) {
-			t.Errorf("[%s] Unexpected amount of lines in audit log: %d", test.desc, len(line))
+		if len(events) != len(test.expected) {
+			t.Errorf("[%s] Unexpected amount of lines in audit log: %d", test.desc, len(events))
```
There is a continue following.

CaoShuFeng replied:
After using t.Run(), continue can't be used.
Changed to t.Fatal().

CaoShuFeng commented:
/test pull-kubernetes-kubemark-e2e-gce

CaoShuFeng commented:
/test pull-kubernetes-kubemark-e2e-gce

CaoShuFeng commented:
/test pull-kubernetes-kubemark-e2e-gce

CaoShuFeng commented:
/test pull-kubernetes-kubemark-e2e-gce

CaoShuFeng commented:
/test pull-kubernetes-kubemark-e2e-gce
/test pull-kubernetes-e2e-gce-etcd3

CaoShuFeng commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit_test.go:
```diff
@@ -855,82 +589,69 @@ func TestAuditJson(t *testing.T) {
+				}
+				if (event.ResponseStatus == nil) != (expect.ResponseStatus == nil) {
+					t.Errorf("Unexpected ResponseStatus: %v", event.ResponseStatus)
+					continue
```
> But this particular place has continue afterwards :)

Do you mean the `continue` in this line?
I think the `continue` here is OK.
For example, there are 3 events in test.expected totally. And it fails in line 646 for the first event, so we run `continue`. And the check in line 650 is skipped, and then we will check the next 2 events.

CaoShuFeng replied:
If we use t.Fatalf(), we will not check other events after the first fatal.
That's my opinion.

crassirostris replied:
Ah, sure, agree, it's fine here, sorry :)

crassirostris commented:
/retest

CaoShuFeng commented:
/test pull-kubernetes-kubemark-e2e-gce

soltysh commented:
/test pull-kubernetes-bazel-build
/test pull-kubernetes-bazel-test

CaoShuFeng commented:
/test pull-kubernetes-e2e-gce-etcd3

soltysh commented:
@calebmiles why did you remove the milestone from here, this is improving the test so it falls under the category of bugfixes/stabilization imo

CaoShuFeng commented:
> @calebmiles why did you remove the milestone from here, this is improving the test so it falls under the category of bugfixes/stabilization imo

Another people is pinged.
s/calebmiles/calebamiles
@calebamiles

soltysh commented:
@jdumars I'm adding a 1.8 milestone back here, since this is fixing tests/improving stability of the code. The only reason it wasn't approved and lgtm-ed on Fri is it needed a rebase. 

/lgtm

CaoShuFeng committed:
enhance unit tests of advance audit feature

This change does three things:
    1. use auditinternal for unit test in filter stage
    2. add a seperate unit test for Audit-ID http header
    3. add unit test for audit log backend

soltysh commented:
/lgtm

soltysh commented:
@sttts for approval

sttts commented:
/approve no-issue
